### DISCUSSION TITLE
Bluetooth Audio

### DISCUSSION THREAD

**Discussion by digitxo** [roc-streaming/roc-toolkit#294] [Discuss]
Does this project work with Bluetooth Audio transmissions?

**Comment by gavv** [comment]
Hi,

No, Roc doesn't support Bluetooth. So far we have plans only on IP-based protocols.

**Comment by eljima** [comment]
I could write the Bluetooth driver(to drive through). I can access some peripherals. If you are okay with this, I will compare the various devices and find the most suitable and practical method to work on it.

**Comment by digitxo** [comment]
> I could write the Bluetooth driver(to drive through). I can access some peripherals. If you are okay with this, I will compare the various devices and find the most suitable and practical method to work on it.

I would love to see real-time Bluetooth audio. Keep me posted!

**Comment by eljima** [comment]
It probably won't be feasible to support all devices but I can work on your one. Post me your device info and maybe I can work on a specific patch for your hardware. Have a nice day

**Comment by digitxo** [comment]
Raspery Pi Zero

**Comment by digitxo** [comment]
> It probably won't be feasible to support all devices but I can work on your one. Post me your device info and maybe I can work on a specific patch for your hardware. Have a nice day

Could you do this with a RPI Zero?

**Comment by eljima** [comment]
Yes, I can. If you want it, I can teach you how to do it and you will program it yourself  learning in the process. I don't ask anything in exchange. Teaching is something I like. Contact me at eljima  outlook   com

If you don't want, then I will make it nonetheless. No mentoring and I will proceed to work on making it.

If you accept we will do it together

**Comment by gavv** [comment]
Hi,

First, regarding to Bluetooth support in Roc.

I can imagine two ways how it could be used:

1. The user may want to read from Roc receiver and write to Bluetooth device, or to read from Bluetooth device and to write to Roc sender. In other words, Roc should allow to connect network Roc sender or receiver to a local Bluetooth device.

2. The user may want to use Roc receiver or sender API to perform actual transfer to or from Bluetooth device. In other words, Roc should allow to create Bluetooth receivers and sender, in addition to network receivers and senders.

Case (1) is, actually, already possible to implement outside of Roc, using current C API. Roc doesn't bother from where or to where you transfer the sound. Asking Roc to do it for you out of the box is a matter of convenience. I'm thinking about adding an API to connect Roc sender / receiver to local audio devices, so that the user will not have to transfer the sound manually. If we'll add such an API, we can add support for Bluetooth devices as well. By the way, if you're using PulseAudio, Roc command-line tools (Roc-recv and Roc-send) will happily work with Bluetooth devices right now. This feature is just not exported to the C API.

Case (2) is something different. It's about using the same Roc API for both network and Bluetooth senders and receivers. But note that it's only about API. The implementation and protocols will be different (though similar). Bluetooth defines its own protocol stack and we won't be able to reuse our network code. We won't be able to use FECFRAME, for example. So it's more about providing an abstraction layer that hides difference between Bluetooth and network. But I'm not sure it's really a good idea. Bluetooth has a lot of specific features (roles, profiles, pairing, ...) that can not be hidden from user. So we may end up with quite different configurations for network and Bluetooth senders / receivers, likely making the idea of the unified API meaningless.

Second, regarding to the implementation.

Creating Bluetooth stack is a huge and complex task[1]. On Linux, it involves kernel drivers, Bluez daemon, oFono, and PulseAudio, and plus some tooling (e.g. bluetoothctl and hcitool). Other platforms have other implementations. I see no reason to re-implement all these code and even don't think it's realistic.

So, taking the above into account, the only feature that may be really useful in Roc, is to allow connecting network Roc sender or receiver to a local audio devices, including Bluetooth devices. On Linux we can just use PulseAudio, and will get Bluetooth support automatically. On other platforms we may need to use some platform-specific API for Bluetooth.

[1] https://software.intel.com/en-us/node/721510

**Comment by digitxo** [comment]
> Yes, I can. If you want it, I can teach you how to do it and you will program it yourself learning in the process. I don't ask anything in exchange. Teaching is something I like. Contact me at eljima outlook com
> 
> If you don't want, then I will make it nonetheless. No mentoring and I will proceed to work on making it.
> 
> If you accept we will do it together

Hi. I didn't catch you on Skype and you're not responding to my emails. Wondering if you aren't noticing them so I posted here. Please get ahold of me I'd really like to work on this project with you.

**Comment by digitxo** [comment]
> Yes, I can. If you want it, I can teach you how to do it and you will program it yourself learning in the process. I don't ask anything in exchange. Teaching is something I like. Contact me at eljima outlook com
> 
> If you don't want, then I will make it nonetheless. No mentoring and I will proceed to work on making it.
> 
> If you accept we will do it together

You there?

**Comment by gavv** [comment]
I think we can close this. Feel free to reopen if necessary.
